### PR TITLE
Week 4 Assessment

### PR DESCRIPTION
**Pull request by tabithatrubey**
# Submitting Week 4 Assessment

### ✅ Completed Files
Mark the following as confirmation that the files in this repository are complete.
- [✅  ] JavaScript Coding Challenges
- [ ✅ ] Ruby Coding Challenges
- [✅  ] Interview Questions

### 📝 Feedback
If you would like specific feedback on a particular problem please leave details here:

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -11,40 +11,86 @@
...
+        
+    })
+})
+// ReferenceError: duplicateAnswers is not defined
+// Test Suites: 1 failed, 1 total
+// Tests:       1 failed, 1 passed, 2 total
+
+// Create a function "duplicateAnswers", that takes in two parameters: 2 arrays
+const duplicateAnswers = (array1, array2) => {
+// We want to combine the two arrays without duplicate values, therefore, we use the spread operator with new Set to add the arrays to a new "set" object & 'set' will not allow for any duplicate values in an array only unique ones, so it's a win win situation.
```
Comment (elysemontano) [inline code-challenges.test.js:60]: Excellent work!

File: code-challenges.test.js [hunk 2]
```diff
@@ -11,40 +11,86 @@
...
+        // expect(shuffleArray(colors1)).toEqual(["yellow", "blue", "pink", "green"])
+        // expect(shuffleArray(colors2)).toEqual(["saffron", "aquamarine", "periwinkle", "indigo", "ochre"])
+        // expect(["purple", "blue", "green", "yellow", "pink"]).toEqual(expect.shuffleArray(colors1))
+        // expect(["saffron", "aquamarine", "periwinkle", "indigo", "ochre"]).toEqual(expect.shuffleArray(colors2))
+        expect(["purple", "blue", "green", "yellow", "pink"]).not.toEqual(shuffleArray(colors1))
+        expect(["chartreuse", "indigo", "periwinkle", "ochre", "aquamarine", "saffron"]).not.toEqual(shuffleArray(colors2))
+    })
+})
+// ReferenceError: shuffleContent is not a function
+// OH MY LANTA THIS JEST TEST WAS SO HARD!!! I didn't even ignore the hyperlink, and somehow still struggled with the test.
```
Comment (elysemontano) [inline code-challenges.test.js:88]: This one is definitely tricky.  While the test is checking that there is some kind of modification to the array, it is not checking that there are certain values which is also important.  A combination of what you have and using the expect.arrayContaining could be really efficient in testing this problem.

File: interview-questions.md [hunk 3]
```diff
@@ -8,42 +8,50 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
-Researched answer:
+Researched answer: Object-oriented programming is one approach to dealing with large and complex software systems. Object-oritented programming is intended to reduce complex software systems into simpler and dynamic classes and objects. Object-oriented programming has four pillars to write clean code: abstraction, encapsulation, inheritance, and polymorphism. Object-oriented programming is reusable, and protects information via encapsulation. Functional programs fixate on logic. The main focus of functional programming is solving a problem. Functions are the star of functional programming as they can serve as arguments to other functions, attach to names, or even just complete a task. A MASSIVE difference between object-oriented and function focused programming that I should have pointed out in my answer above, is the difference between error messages. Personally, I think error messages in object-oriented programs, like Ruby, are much more user-friendly and more explicit regarding the error, how to fix it, and where to find it. However, in a functional focused program, like JavaScript, error messages can be a bit ambiguous.
```
Comment (elysemontano) [inline interview-questions.md:13]: Outstanding answer!

File: interview-questions.md [hunk 4]
```diff
@@ -8,42 +8,50 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+                                                                p roll_call 
+                                                                #Output: "Tabi is present" 
 
-Researched answer:
+Researched answer: Ruby does not require a 'return' like JavaScript does in every circumstance. Therefore, when 'return' is not written, Ruby returns an implicit return which means the value of the last line will be returned. 
 
 4. What is a block in Ruby?
 
-Your answer:
+Your answer: A block in Ruby is the portion that holds the logic in a method. A block in Ruby is defined by 'do' at the beginning of the logic and 'end' to indicate the logic has concluded. A block is traditionally followed by pipes || that pass the parameter(s) of the method, and can allow for dynamic code to be ran properly.
```
Comment (elysemontano) [inline interview-questions.md:33]: I imagine you put your hands up when writing |pipes| in this answer 😆

File: ruby-challenges.rb [hunk 5]
```diff
@@ -2,15 +2,25 @@
...
+# I kept getting stuck becuase I put 'p #{num}  is even"' and 'p #{num}  is odd' when that line isn't supposed to have a print.
 num1 = 7
 # Expected output: '7 is odd'
 num2 = 42
 # Expected output: '42 is even'
 num3 = 221
 # Expected output: '221 is odd'
 
+def even_odd num
+    if num % 2 == 0
```
Comment (elysemontano) [inline ruby-challenges.rb:14]: To dig into the expansive built in methods available in Ruby, check out the .even? or .odd? method.

File: ruby-challenges.rb [hunk 6]
```diff
@@ -22,13 +32,33 @@
...
+# -------------------3) Create a method that takes in a string and checks if the string is a palindrome. A palindrome is the same word spelled forward or backward. Use the test variables provided.
+# Was so proud of myself for completely this one so quickly, and then my tester2 kept returning as a palindrome. Forgot to add loose equality!
 palindrome_tester1 = 'Racecar'
 # Expected output: 'Racecar is a palindrome'
 palindrome_tester2 = 'LEARN'
 # Expected output: 'LEARN is not a palindrome'
 palindrome_tester3 = 'Rotator'
 # Expected output: 'Rotator is a palindrome'
+
+def palindrome string
```
Comment (elysemontano) [inline ruby-challenges.rb:53]: Excellent work in this section!

**Review by elysemontano — APPROVED** [review]
Crushed this assessment Tabi!  You definitely stepped up a notch in trying new approaches.   I left some feedback throughout for you to take a look at. 
Current assessment level is advanced beginner, however, you are really approaching competent with the quality of this assessment.  Try playing around with the many different Ruby methods available and consider what your tests are testing.  Consider edge cases when possible.  Overall though, excellent work!!
